sql/table.go

```diff
@@ -295,26 +295,34 @@ type indexEntry struct {
 }
 
 func encodeSecondaryIndexes(tableID ID, indexes []IndexDescriptor,
-	colMap map[ColumnID]int, values []parser.Datum, primaryIndexKeySuffix []byte) ([]indexEntry, error) {
+	colMap map[ColumnID]int, values []parser.Datum) ([]indexEntry, error) {
 	var secondaryIndexEntries []indexEntry
 	for _, secondaryIndex := range indexes {
 		secondaryIndexKeyPrefix := MakeIndexKeyPrefix(tableID, secondaryIndex.ID)
-		secondaryIndexKey, containsNull, err := encodeIndexKey(secondaryIndex.ColumnIDs, colMap, values, secondaryIndexKeyPrefix)
+		secondaryIndexKey, containsNull, err := encodeIndexKey(
+			secondaryIndex.ColumnIDs, colMap, values, secondaryIndexKeyPrefix)
 		if err != nil {
 			return nil, err
 		}
 
-		if secondaryIndex.Unique && !containsNull {
-			secondaryIndexEntries = append(secondaryIndexEntries, indexEntry{
-				key:   secondaryIndexKey,
-				value: primaryIndexKeySuffix,
-			})
-		} else {
-			secondaryIndexEntries = append(secondaryIndexEntries, indexEntry{
-				key:   append(secondaryIndexKey, primaryIndexKeySuffix...),
-				value: nil,
-			})
+		extraKey, _, err := encodeIndexKey(secondaryIndex.ExtraColumnIDs, colMap, values, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		var entry indexEntry
+		entry.key = secondaryIndexKey
+
+		if !secondaryIndex.Unique || containsNull {
+			// If the index is not unique or it contains a NULL value, append
+			// extraKey to the key in order to make it unique.
+			entry.key = append(entry.key, extraKey...)
 		}
+		if secondaryIndex.Unique {
+			entry.value = extraKey
+		}
```

Comment:
Yes, this is intentional. Unique secondary indexes always have `extraKeys` stored in the value so that scanning can know to find them there. And `extraKeys` is appended to the key when the unique index key contains a NULL value in order to make the key unique. We could try to optimize this a bit further, but it seems rare and not worth doing.